frame/header-mmr/src/lib.rs

```diff
@@ -176,61 +227,18 @@ pub mod pallet {
 	}
 	impl<T: Config> MMRT<T::BlockNumber, T::Hash> for Pallet<T> {
 		fn get_root(block_number: T::BlockNumber) -> Option<T::Hash> {
-			let store = <ModuleMMRStore<T>>::default();
-			let mmr_size = mmr::leaf_index_to_mmr_size(block_number.saturated_into::<u64>() as _);
-			let mmr = <MMR<_, MMRMerge<T>, _>>::new(mmr_size, store);
-
-			if let Ok(mmr_root) = mmr.get_root() {
-				Some(mmr_root)
-			} else {
-				None
-			}
-		}
-	}
-
-	pub struct ModuleMMRStore<T>(PhantomData<T>);
-	impl<T> Default for ModuleMMRStore<T> {
-		fn default() -> Self {
-			ModuleMMRStore(sp_std::marker::PhantomData)
-		}
-	}
-	impl<T: Config> MMRStore<T::Hash> for ModuleMMRStore<T> {
-		fn get_elem(&self, pos: u64) -> MMRResult<Option<T::Hash>> {
-			Ok(<Pallet<T>>::mmr_node_list(pos))
-		}
+			let size = mmr::leaf_index_to_mmr_size(block_number.saturated_into());
 
-		fn append(&mut self, pos: u64, elems: Vec<T::Hash>) -> MMRResult<()> {
-			let mmr_count = <MMRCounter<T>>::get();
-			if pos != mmr_count {
-				// Must be append only.
-				Err(Error::InconsistentStore)?;
-			}
-			let elems_len = elems.len() as u64;
-
-			for (i, elem) in elems.into_iter().enumerate() {
-				<MMRNodeList<T>>::insert(mmr_count + i as u64, elem);
-			}
-
-			// increment counter
-			<MMRCounter<T>>::put(mmr_count + elems_len);
-
-			Ok(())
-		}
-	}
-
-	pub struct MMRMerge<T>(PhantomData<T>);
-	impl<T: Config> Merge for MMRMerge<T> {
-		type Item = <T as frame_system::Config>::Hash;
-
-		fn merge(lhs: &Self::Item, rhs: &Self::Item) -> Self::Item {
-			let encodable = (lhs, rhs);
-			<T as frame_system::Config>::Hashing::hash_of(&encodable)
+			<Mmr<RuntimeStorage, T>>::with_size(size).get_root().ok()
```

Comment:
`<Mmr<RuntimeStorage, T>>` will not be able to get root for block_numbers expect the last one.

Reply:
This method can be removed if not used in runtime.

Reply:
`relay-authorities` require this method. I'm considering if I should change this to `OffchainStorage`.

Reply:
Since `OffchainStorage` can not be used in runtime (execute by each node including full node and non-indexing node), we have to remove this.

For the `submit_signed_mmr_root` dispatch call in `relay-authorities`, since the mmr_root for history block can not be calculated on chain , the call will require relayers to submit it together with the signature, will change following api and add an extra parameter `mmr_root`.
```
pub fn submit_signed_mmr_root(
			origin,
			block_number: BlockNumberFor<T>,
			signature: RelayAuthoritySignature<T, I>
		)
```

This breaks api, and need companion change in bridger, cc @wuminzhe

Reply:
I got a new idea.

We could store the mmr root while scheduled. https://github.com/darwinia-network/darwinia-common/blob/master/frame/bridge/relay-authorities/src/lib.rs#L857-L872

And this won't break the API.

Reply:
> We could store the mmr root (the latest one calculated from RuntimeStorage) while scheduled. https://github.com/darwinia-network/darwinia-common/blob/master/frame/bridge/relay-authorities/src/lib.rs#L857-L872

This is a better way, first schedule the block_number need to get the mmr root, and get the mmr root on the right block's on_initialize.

Change the original `fn get_root(block_number: T::BlockNumber)` to `fn get_mmr_root()`, it is from genesis to parent.

Be careful with the right block number to call `get_mmr_root` (before header_mmr:on_finalize, or behind)

Reply:
Orignal `fn get_root(block_number: T::BlockNumber)` is [0..= block_number], header_hash(0) ..= header_hash(block_number), inclusive, the first leave is generate by block(1)'s on_finalize().

![image](https://user-images.githubusercontent.com/1070122/123214161-73391c80-d4f9-11eb-989e-d874482738ae.png)


so `get_root(block_number)` should not be called on (block_number + 2)'s on_initialize()?